09_json/nickolee/cmd/puppy-server/main.go

```diff
@@ -0,0 +1,84 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+
+	puppy "github.com/anz-bank/go-course/09_json/nickolee/pkg/puppy"
+	store "github.com/anz-bank/go-course/09_json/nickolee/pkg/puppy/store"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+var out io.Writer = os.Stdout
+
+var (
+	args     = os.Args[1:]
+	fileName = kingpin.Flag("data", "This flag supplies a puppy JSON file name").Short('d').String()
+)
+
+func main() {
+	// parse flags and read json from file
+	fileData := parseFlagsAndLoadFile() // fileData var to store raw bytes read in from file
+
+	// parse/unmarshal the json and we now have []puppy (Go objects)
+	puppies := unmarshalPuppies(fileData)
+
+	// instantiate new puppy store
+	store := store.NewMapStore()
+
+	// create some puppies loaded in from json file and store in store (no pun intended lol)
+	for _, pup := range puppies {
+		pup := pup // to avoid scopelint error: Using a reference for the variable on range scope `pup`
+		createdPuppy, _ := store.CreatePuppy(&pup)
+		fmt.Fprintf(out, "Puppy with ID %d has been created\n", createdPuppy)
+		retrievedPuppy, _ := store.ReadPuppy(createdPuppy)
+		fmt.Fprintln(out, "Retrieved puppy:", retrievedPuppy)
+	}
+}
+
+// parse flags and load file
+func parseFlagsAndLoadFile() (fileData []byte) {
+	// parse cli flags and check for flag parse error
+	if _, parseError := kingpin.CommandLine.Parse(args); parseError != nil {
+		printUsage()
+		panic(parseError) // this also includes a panic
+	}
+
+	// if data flag is supplied. Reading data from file and check for file argument error
+	f, err := os.Open(*fileName) // At this point Go knows about the file but not what's in it
+	if err != nil {
+		printUsage()
+		panic(err)
+	}
```

Comment:
`defer f.Close()` to make sure the file is closed after this function returns (you don't need this file open after this function returns)

Reply:
Thanks man. Done. Although is this just for good practice? Cos in my case I don't think it matters if the file is still open?

Reply:
Mainly good practice, should get in the habit of closing resources as soon as you are done with them.